### ISSUE TITLE
set_random_seed does not make temperature based decoding deterministic

### ISSUE THREAD

**Issue by ozancaglayan** [OpenNMT/CTranslate2#1757]
Hello,

I'm experimenting with the `set_random_seed()` method of CTranslate2 (through faster whisper). Without temperature based decoding, the results from Whisper ASR models are always the same on CPUs and GPUs for the same file, e.g. I don't have any other source of stochasticity.

Previously, I was calling the `set_random_seed()` only at model creation. Now I changed the logic to set it to the same value before every `generate()` call. But still, I can't make the sampling deterministic. I tried on CPU and GPU, both give different results.

Any idea what am I missing? Thanks

**Comment by minhthuc2502** [comment]
Hello,

With the temperature set, you are changing the probability distribution by making it either sharper or flatter. We have a softmax after multiplying score with the temperature. Even with the same seed, if the distribution is very flat, it could lead to the variety in the result (maybe precision error of fp). Try to set a temperature < 1 and see if the result is deterministic.

**Comment by ozancaglayan** [comment]
Hi,

Thanks! I was actually trying with temperatures of 0.2, 0.4, 0.8 etc.

**Comment by UpiterSoft** [comment]
Code to reproduce the problem
`
import json
import textwrap
import numpy as np

import ctranslate2
from transformers import AutoTokenizer, AutoModelForCausalLM

import torch

def generate_ct2(generator, tokenizer, tokens, num_hypotheses=3):
    result = generator.generate_batch(
        start_tokens=tokens * num_hypotheses,
        max_length=8,
        sampling_temperature=1.0,
        sampling_topk=10,
        sampling_topp=0.95,
        num_hypotheses=num_hypotheses,
        beam_size=1,
        include_prompt_in_result=False,
        end_token=[tokenizer.convert_ids_to_tokens(tokenizer.encode(text))[0]
                   for text in ['\n', '.', '<|endoftext|>']],
    )

    scores = generator.score_batch([tokens[0] + s for s in result[0].sequences])
    probs = [np.exp(np.array(s.log_probs[len(tokens[0]) - 1:])) for s in scores]

    texts = tokenizer.batch_decode(
        result[0].sequences_ids,
    )
    return texts, probs


def generate_transformers(model, tokenizer, query, num_hypotheses=3):
    tokens = tokenizer(query, return_tensors='pt', padding=False).to('cuda')
    outputs = model.generate(**tokens,
                             num_return_sequences=num_hypotheses,
                             max_new_tokens=8,
                             temperature=1.0,
                             top_k=10,
                             top_p=0.95,
                             do_sample=True,
                             tokenizer=tokenizer,
                             )
    texts = tokenizer.batch_decode(outputs[:, tokens.input_ids.shape[1]:], skip_special_tokens=True)
    return texts


def main():
    # > ct2-transformers-converter --model gpt2 --quantization float16 --output_dir=./gpt2-ct2
    model_path_ct2 = './gpt2-ct2'
    tokenizer_name = 'gpt2'
    model_name = 'gpt2'

    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)

    query = textwrap.dedent("""
        Q: Pick a number in range 1 0 0 0.. 9 9 9 9
        A: I pick number 3 9 7 2
        Q: Pick a number in range 1 0 0 0.. 9 9 9 9
        A: I pick number 8 6 0 4
        Q: Pick a number in range 1 0 0 0.. 9 9 9 9
        A: I pick number
    """).rstrip()
    tokens = [tokenizer.convert_ids_to_tokens(tokenizer.encode(query))]

    generator = ctranslate2.Generator(model_path_ct2,
                                      device="cuda",
                                      # compute_type='float32',  # does not make any difference
                                      # intra_threads=1,  # does not make any difference
                                      # inter_threads=1,  # does not make any difference
                                      # max_queued_batches=1,  # does not make any difference
                                      )

    model = AutoModelForCausalLM.from_pretrained(model_name).cuda()
    model.generation_config.pad_token_id = model.generation_config.eos_token_id
    model.generation_config.stop_strings = ['\n', '.', '<|endoftext|>']

    print('CTranslate2')
    for seed in [42, 42, 42, 43, 44]:
        # torch.cuda.random.manual_seed_all(seed)
        ctranslate2.set_random_seed(seed)

        texts_ct2, probs = generate_ct2(generator, tokenizer, tokens, num_hypotheses=3)

        for text, probs in zip(texts_ct2, probs):
            print(f'seed={seed}, {json.dumps(text):<12}: {np.round(probs, 3)}')

        print()

    print('Transformers')
    for seed in [42, 42, 42, 43, 44]:
        torch.cuda.random.manual_seed_all(seed)

        texts_transformers = generate_transformers(model, tokenizer, query, num_hypotheses=3)

        for text in texts_transformers:
            print(f'seed={seed}, {json.dumps(text)}')

        print()

    {}


if __name__ == '__main__':
    main()
`

Output: `
**CTranslate2**
seed=42, " 8 6 0 4"  : [0.069 0.53  0.96  0.989]
seed=42, " 9 5 0 5"  : [0.273 0.287 0.86  0.099]
seed=42, " 10 4 0 2" : [0.101 0.164 0.784 0.142]

seed=42, " 9 5 0 3"  : [0.273 0.287 0.86  0.237]
seed=42, " 6 7 0 4"  : [0.042 0.073 0.793 0.748]
seed=42, " 7 5 0 5"  : [0.054 0.326 0.878 0.089]

seed=42, " 9 5 0 3"  : [0.273 0.287 0.86  0.237]
seed=42, " 8 5 0 4"  : [0.069 0.195 0.93  0.588]
seed=42, " 10 5 0 6" : [0.101 0.505 0.872 0.056]

seed=43, " 9 6 0 4"  : [0.273 0.224 0.816 0.907]
seed=43, " 12 6 0 4" : [0.037 0.217 0.779 0.706]
seed=43, " 9 5 6 2"  : [0.273 0.287 0.014 0.133]

seed=44, " 8 6 0 4"  : [0.069 0.53  0.96  0.989]
seed=44, " 9 7 0 4"  : [0.273 0.093 0.497 0.577]
seed=44, " 12 10 3 4": [0.037 0.07  0.063 0.084]

**Transformers**
seed=42, " 8 6 0 4\n"
seed=42, " 8 6 0 4\n"
seed=42, " 9 4 0 4\n"

seed=42, " 8 6 0 4\n"
seed=42, " 8 6 0 4\n"
seed=42, " 9 4 0 4\n"

seed=42, " 8 6 0 4\n"
seed=42, " 8 6 0 4\n"
seed=42, " 9 4 0 4\n"

seed=43, " 3 6 0 4\n"
seed=43, " 1 8 0 0\n"
seed=43, " 1 7 3 2\n"

seed=44, " 7 6 0 4\n"
seed=44, " 11 10 0 3\n"
seed=44, " 7 5 0 4\n"
`

CTranslate2 with seed=42 is expected to generate equal responses, as it happens for Transformers

Name: ctranslate2
Version: 4.5.0

**Comment by UpiterSoft** [comment]
@minhthuc2502 can you please take a look? there is a code to reproduce problem
